transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringEventLoop.java

```diff
@@ -0,0 +1,157 @@
+                                       0,
+                                       0,
+                                       0);
+                        addNewEvent(event);
+                        ringBuffer.getIoUringSubmissionQueue().submit();
+                        break;
+                    case READ:
+                        System.out.println("Eventlloop Read Res: " + ioUringCqe.getRes());
+                        System.out.println("Eventloop Fd: " + event.getAbstractIOUringChannel().getSocket().getFd());
+                        ByteBuf byteBuf = event.getReadBuffer();
+                        int localReadAmount = ioUringCqe.getRes();
+                        if (localReadAmount > 0) {
+                            byteBuf.writerIndex(byteBuf.writerIndex() + localReadAmount);
+                        }
+
+                        final IOUringRecvByteAllocatorHandle allocHandle =
+                                (IOUringRecvByteAllocatorHandle) event.getAbstractIOUringChannel().unsafe()
+                                                                      .recvBufAllocHandle();
+                        final ChannelPipeline pipeline = event.getAbstractIOUringChannel().pipeline();
+
+                        allocHandle.lastBytesRead(localReadAmount);
+                        if (allocHandle.lastBytesRead() <= 0) {
+                            // nothing was read, release the buffer.
+                            byteBuf.release();
+                            byteBuf = null;
+                            break;
+                        }
+
+                        allocHandle.incMessagesRead(1);
+                        //readPending = false;
+                        pipeline.fireChannelRead(byteBuf);
+                        byteBuf = null;
+                        allocHandle.readComplete();
+                        pipeline.fireChannelReadComplete();
+                        event.getAbstractIOUringChannel().executeReadEvent();
+                        break;
+                    case WRITE:
+                        System.out.println("Eventloop Write Res: " + ioUringCqe.getRes());
+                        System.out.println("Eventloop Fd: " + event.getAbstractIOUringChannel().getSocket().getFd());
+                        //remove bytes
+                        int localFlushAmount = ioUringCqe.getRes();
+                        if (localFlushAmount > 0) {
+                            event.getAbstractIOUringChannel().unsafe().outboundBuffer().removeBytes(localFlushAmount);
+                        }
+                        break;
+                    }
+                } else {
+                    System.out.println("Event is null!!!! ");
+                }
+            }
+            //run tasks
+            if (hasTasks()) {
+                runAllTasks();
+            }
+            try {
+                Thread.sleep(10);
```

Comment:
@normanmaurer it's temporary, I don't quite understand how that works with the wakeups in the eventloop

Reply:
yeah I would also prefer the second approach, it's probably rare that threads will have to wait for a lock, as we will use lock for each channel which means we move all `Channel` changes to taskQueue something like that `readOperation(Event event)` so we dont need a lock for the completion queue only for the submission queue and if the user write something( which is a EventLoop Thread but I'm not sure)

Reply:
I think the ChannelPipeline is shared between the Channels, then the approach lock for each channel would not work...
or is ChannelPipeLine thread safe?

Reply:
I don't think I understand what you mean by "move all `Channel` changes to taskQueue", or why the channels need to be locked.

Reply:
@chrisvest My approach would be that the completionQueue thread(io_uring_enter) get the event object via Hashmap which is then added to the `taskQueue`
For example(pseudo code):
```
        public Runnable readEvent(final Event event) {
            final int test = 2233;
            final Runnable readRunnable = new Runnable() {

                @Override
                public void run() {

                    ByteBuf byteBuf = event.getReadBuffer();
                    int localReadAmount = event.getRes();
                    if (localReadAmount > 0) {
                        byteBuf.writerIndex(byteBuf.writerIndex() + localReadAmount);
                    }

                    final IOUringRecvByteAllocatorHandle allocHandle =
                            (IOUringRecvByteAllocatorHandle) event.getAbstractIOUringChannel().unsafe()
                                                                  .recvBufAllocHandle();
                    final ChannelPipeline pipeline = event.getAbstractIOUringChannel().pipeline();

                    allocHandle.lastBytesRead(localReadAmount);
                    if (allocHandle.lastBytesRead() <= 0) {
                        // nothing was read, release the buffer.
                        byteBuf.release();
                        byteBuf = null;
                        break;
                    }

                    allocHandle.incMessagesRead(1);
                    //readPending = false;
                    pipeline.fireChannelRead(byteBuf);
                }
            };


            return readRunnable;
        }
```
I think we dont need a lock for the read and accept event because only one thread submission queue will modify the channel data
https://github.com/1Jo1/netty/blob/io-uring/transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringEventLoop.java#L156
https://github.com/1Jo1/netty/blob/io-uring/transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringSocketChannel.java#L60
  
But the user and the submission queue thread will modify the same `ChannelOutboundBuffer` instance so we need a lock the for the write event
https://github.com/1Jo1/netty/blob/io-uring/transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringEventLoop.java#L171
https://github.com/1Jo1/netty/blob/io-uring/transport-native-io_uring/src/main/java/io/netty/channel/uring/AbstractIOUringChannel.java#L173
does that make sense?

Reply:
Okay, right… I'm not sure how that `ChannelOutboundBuffer` is normally handled, but it looks like a single-threaded object. It sounds correct, then, to move the completion processing back onto the task queue like you describe. Looks like epoll, kqueue, and NIO, avoid the lock by using blocking writes, if I'm reading it correctly.

Reply:
@chrisvest yeah exactly :) I'm not sure which approach is more efficient: avoid the lock by using blocking writes or use a lock?

Reply:
I guess it's easier to implement the `avoid the lock` approach